build-changed-pages.py

```diff
@@ -0,0 +1,242 @@
+        except FileNotFoundError:
+            print(f"{assembly_path} is active in at least one topic map, but the the file is not found.")
+            sys.exit(1)
+    # Sort and deduplicate the module list entries:
+    module_list.sort()
+    module_list = list(dict.fromkeys(module_list))
+
+# Get the relative path for all active snippets referenced in the active assembly files:
+def list_active_snippet_files():
+    global snippet_list
+    snippet_list = []
+    for a in assembly_list:
+        assembly_path = clone_dir + "/" + a
+        try:
+            with open(assembly_path) as f:
+                for line in f:
+                    if line.startswith("include::") and "snippets/" in line:
+                        snippet_path = re.search('snippets/[^"\r\n]*.adoc', line)[0]
+                        snippet_list.append(snippet_path)
+        except FileNotFoundError:
+            print(f"{assembly_path} is active in at least one topic map, but the the file is not found.")
+            sys.exit(1)
+    # Sort and deduplicate the module list entries:
+    snippet_list.sort()
+    snippet_list = list(dict.fromkeys(snippet_list))
+
+# Create a list of pages that are changed by the latest commit in the specified branch. Then, build previews only for those changed pages:
+def list_changed_files_in_pr():
+    # Create an empty pages to build list:
+    pages_to_build = []
+    # Create a dictionary of files changed in the HEAD commit for the current branch. In the dictionary, each key is the changed file name and the value is the change type:
+    head_commit = repo.head.commit
+    for i in head_commit.diff("HEAD~1"):
+        d.update({i.a_path: i.change_type})
+    # For each changed file in HEAD, check if the files are active and compile a list of pages to build:
+    for key, value in d.items():
+        # If a changed file is an assembly file that is active in the topic map, add it to the list of pages to build:
+        if ".adoc" in key and key in assembly_list and "D" not in value:
+            pages_to_build.append(key.replace('.adoc',''))
+        # Check if each changed module file is in an active assembly:
+        if ".adoc" in key and key in module_list and "D" not in value:
+            include_string = "include::" + key + "["
+            for i in assembly_list:
+                with open(clone_dir + "/" + i, 'r') as file:
+                    content = file.read()
+                    if key in content:
+                        pages_to_build.append(i.replace('.adoc',''))
+    # Sort and deduplicate the list of pages that are changed in the HEAD commit:
+    pages_to_build.sort()
+    pages_to_build = list(dict.fromkeys(pages_to_build))
+    # Run 'asciibinder build' for each page that is changed in the HEAD commit, including any unchanged assemblies that include changed modules:
+    print(f"Attempting to build the pages changed in the HEAD commit:")
+    for i in pages_to_build:
```

Comment:
Could these processes be threaded in parallel? https://stackoverflow.com/a/14533902/7010427